hbase-client/src/main/java/org/apache/hadoop/hbase/client/AsyncBufferedMutatorImpl.java

```diff
@@ -153,24 +220,45 @@ Stream.<CompletableFuture<Void>> generate(CompletableFuture::new).limit(mutation
       bufferedSize += heapSize;
       if (bufferedSize >= writeBufferSize) {
         LOG.trace("Flushing because write buffer size {} reached", writeBufferSize);
-        internalFlush();
-      } else if (maxMutations > 0 && this.mutations.size() >= maxMutations) {
+        // drain now and send after releasing the lock
+        batch = drainBatch();
+      } else if (this.mutations.size() >= maxMutations) {
```

Comment:
did we lose the positive `maxMutations` check? (Looks like default is `-1` if I'm reading code correctly, which would lead to flush on every mutate.)
```suggestion
      } else if (maxMutations > 0 && this.mutations.size() >= maxMutations) {
```

Reply:
It seems we did. I will fix this oversight.